src/irmin-pack/s.ml

```diff
@@ -55,41 +55,72 @@ module type Specifics = sig
   (** [flush t] flush read-write pack on disk. Raises [RO_Not_Allowed] if called
       by a readonly instance.*)
 
-  val start_gc :
-    ?unlink:bool ->
-    throttle:[ `Block | `Skip ] ->
-    repo ->
-    commit_key ->
-    bool Lwt.t
-  (** [start_gc] tries to start the gc process and returns true if the gc is
-      launched.
+  module Gc : sig
+    (** GC *)
+
+    type throttle = [ `Block | `Skip ]
+    (** Type to control how to handle an already running GC process.
+
+        [Block] means to wait until it finishes, and iff it succeeds, start a
+        new GC process
+
+        [Skip] means ignore requests for a new GC process if one is already
+        running *)
+
+    (** {2 Low-level API} *)
+    val start_exn :
+      ?unlink:bool -> throttle:throttle -> repo -> commit_key -> bool Lwt.t
+    (** [start_exn] tries to start the gc process and returns true if the gc is
+        launched.
+
+        If [unlink] is false then temporary files and files from the previous
+        generation will be kept on disk after the gc finished. This option is
+        useful for debugging. The default is true.
 
-      If [unlink] is false then temporary files and files from the previous
-      generation will be kept on disk after the gc finished. This option is
-      useful for debugging. The default is true.
+        If [throttle] is [Skip] and there is a concurrent gc, [start_gc] returns
+        false immediately without launching a second gc. If [throttle] is
+        [Block] and there is a concurrent gc, [start_gc] blocks waiting for the
+        previous gc to finish and only after launches a second gc. If the
+        previous GC failed, the function returns without launching a new GC.
 
-      If [throttle] is [Skip] and there is a concurrent gc, [start_gc] returns
-      false immediately without launching a second gc. If [throttle] is [Block]
-      and there is a concurrent gc, [start_gc] blocks waiting for the previous
-      gc to finish and only after launches a second gc. If the previous GC
-      failed, the function returns without launching a new GC.
+        TODO: Detail exceptions raised. *)
 
-      TODO: Detail exceptions raised. *)
+    val finalise_exn : ?wait:bool -> repo -> bool Lwt.t
+    (** [finalise_exn ?wait repo] waits for the gc process to finish in order to
```

Comment:
I am worried by the following scenario: there is lots of concurrency, and multiple Lwt promises are invoking start_exn with throttle=Block; so presumably each promise then needs to call finalise_gc for the particular GC run that they (eventually) trigger? How do we link the call to start_exn, with the actual start of the relevant GC run, with the call to finalise_exn? (If we think of a GC run as a "resource", we would return a unique handle for each call to start_exn, which we could then poll to see when the GC actually starts running, and finally call finalise_exn on when the GC has completed...)

Reply:
Agreed. I have had a similar concern in the back of my mind while working on this and the example code.

The current tezos integration uses `Block`, which I copied into the `run` function. I think I will change it to `Skip` to at least make the high-level api safer in this regard. (My hope is to change the tezos integration to use the high-level api)

Lets open an issue for a separate PR for the change you're thinking about.

Reply:
Created an issue to remove throttle and only have `Skip` behavior for now. https://github.com/mirage/irmin/issues/1978